### PR TITLE
wallet: fix sendall creates tx that fails tx-size check

### LINKED ISSUE #26011: sendall creates tx that fails tx-size mempool check
Steps to reproduce:
* Create many inputs
* Call `sendall`

Error: `2022-09-05T09:07:57Z [httpworker.1] [wallet/wallet.h:827] [WalletLogPrintf] [default wallet] CommitTransaction(): Transaction cannot be broadcast immediately, tx-size`

### PR DESCRIPTION
**Pull request by kouloumos**
Fixes https://github.com/bitcoin/bitcoin/issues/26011

The `sendall` RPC doesn't use `CreateTransactionInternal` as the rest of
the wallet RPCs. [This has already been discussed in the original PR](https://github.com/bitcoin/bitcoin/pull/24118#issuecomment-1029462114). 
By not going through that path, it never checks the transaction's weight 
against the maximum tx weight for transactions we're willing to relay.
https://github.com/bitcoin/bitcoin/blob/447f50e4aed9a8b1d80e1891cda85801aeb80b4e/src/wallet/spend.cpp#L1013-L1018
This PR adds a check for tx-size as well as test coverage for that case.

_Note: It seems that the test takes a bit of time on slower machines, 
I'm not sure if dropping it might be for the better._

### CODE REVIEW

File: src/wallet/rpc/spend.cpp
Comment (ishaanam) [inline src/wallet/rpc/spend.cpp:1412]: Might this check make more sense if it were placed in `FinishTransaction` (which is called by `send` and `sendall`) within the section that is only performed if the transaction is complete? The advantages of this would be that A) this check is now added for both `sendall` and `send`, and B) this check isn't performed if the transaction doesn't contain a complete set of signatures.

**Review by ishaanam — COMMENTED** [review]
Concept ACK
It may be out of the scope of this PR to fix this, but I wanted to note that it seems like there are a few other checks that `CreateTransactionInternal` performs that aren't implemented in `sendall` but that I think would still be relevant. Namely:

- whether the feerate exceeds the default maximum feerate set by the user
- whether the tx will pass the mempool's chain limits

File: src/wallet/rpc/spend.cpp
Comment (kouloumos) [reply src/wallet/rpc/spend.cpp:1412]: [`send` is using `FundTransaction`](https://github.com/bitcoin/bitcoin/blob/19585eeb778ac093d7b9e93abd099629849c3fef/src/wallet/rpc/spend.cpp#L1228) which then calls the [`wallet::FundTransaction`](https://github.com/bitcoin/bitcoin/blob/19585eeb778ac093d7b9e93abd099629849c3fef/src/wallet/spend.cpp#L1087) which uses `CreateTransaction` therefore performing those checks.

**Comment by kouloumos** [comment]
> * whether the feerate exceeds the default maximum feerate set by the user

There is already [this check](https://github.com/bitcoin/bitcoin/blob/a9049dd296a47319038d8c34bcc1ca46c89dee9c/src/wallet/rpc/spend.cpp#L1358-L1368) against the `fee_rate` passed by the user. I can't find any other relevant check at `CreateTransactionInternal`. Could you give more details on that?

> * whether the tx will pass the mempool's chain limits

My understanding is that this is not a problem because this RPC only uses confirmed UTXOs.

File: src/wallet/rpc/spend.cpp
Comment (ishaanam) [reply src/wallet/rpc/spend.cpp:1412]: Ok, I hadn't noticed that.

**Comment by ishaanam** [comment]
> I can't find any other relevant check at `CreateTransactionInternal`. Could you give more details on that?

I meant [this](https://github.com/bitcoin/bitcoin/blob/a9049dd296a47319038d8c34bcc1ca46c89dee9c/src/wallet/spend.cpp#L1020-L1022) check, which checks that the total fee is not greater than the maximum fee (set using -maxtxfee). Earlier I should have said fee instead of feerate.

**Comment by achow101** [comment]
ACK 8d822033bcdf6888b900d5a65ff21f3d8b9d30d5

**Review by glozow — COMMENTED** [review]
utACK 8d822033bcdf6888b900d5a65ff21f3d8b9d30d5

Agree we should stop sendall from creating a nonstandard tx. I think a more long term solution would be to still enable sendall to do what it's supposed to do, but create multiple transactions.

File: test/functional/wallet_sendall.py
Comment (kouloumos) [reply test/functional/wallet_sendall.py:267]: Do you see this multiple transactions capability to be helpful on any other scenario except of when the wallet has many UTXOs?

File: src/wallet/rpc/spend.cpp
Comment (kouloumos) [reply src/wallet/rpc/spend.cpp]: Done, definitely adds more context.

**Comment by kouloumos** [comment]
> which checks that the total fee is not greater than the maximum fee (set using -maxtxfee). 

Good catch, I see that this change is now https://github.com/bitcoin/bitcoin/pull/26084.

Force pushed only [to add more context to the added check](https://github.com/bitcoin/bitcoin/pull/26024/#discussion_r970651614).

**Comment by achow101** [comment]
ACK 2b4fc04916e3ed43fcf7297ef6d7c9d71148c844

File: src/wallet/rpc/spend.cpp
Comment (kouloumos) [reply src/wallet/rpc/spend.cpp:1411]: I initially tracked what error code the same check was returning from `CreateTransactionInternal`.

I now see that the 2 paths to call `CreateTransaction` return a different error code:

**RPC_WALLET_INSUFFICIENT_FUNDS** < `SendMoney` < `CreateTransaction` < `CreateTransactionInternal`
https://github.com/bitcoin/bitcoin/blob/718304d222671f98d2335cd9b90a3022f62d7b21/src/wallet/rpc/spend.cpp#L159-L162

**RPC_WALLET_ERROR** < `FundTransaction` < `CreateTransaction` < `CreateTransactionInternal`
https://github.com/bitcoin/bitcoin/blob/718304d222671f98d2335cd9b90a3022f62d7b21/src/wallet/rpc/spend.cpp#L702-L704

I am not sure if I am missing something but it seems that there is ambiguity. Probably because `CreateTransactionInternal` includes a variety of checks that in the end need to be represented with a single error code.

For example, a failed fee estimation is returned differently based on the RPC
https://github.com/bitcoin/bitcoin/blob/718304d222671f98d2335cd9b90a3022f62d7b21/test/functional/wallet_fallbackfee.py#L27-L29

Nevertheless, I agree with you that in this case RPC_WALLET_ERROR is more appropriate. I will change it together with the needed rebase.

File: test/functional/wallet_sendall.py
Comment (glozow) [reply test/functional/wallet_sendall.py:267]: No. But it seems quite limiting to just not be able to sendall when the wallet gets bigger. I might be wrong but isn't consolidation one of its use cases? And it's a bit ugly to not be able to use cleanup after this test. Anyway I don't want to hold up this PR, so feel free to resolve.

Comment (kouloumos) [reply test/functional/wallet_sendall.py:267]: It was more of a curiosity question. Your argument seems valid, as per the current functional test, you need ~1600 inputs to reach the limit.

**Review by w0xlt — APPROVED** [review]
ACK https://github.com/bitcoin/bitcoin/pull/26024/commits/2b4fc04916e3ed43fcf7297ef6d7c9d71148c844

**Review by w0xlt — APPROVED** [review]
reACK https://github.com/bitcoin/bitcoin/pull/26024/commits/cc434cbf583ec8d1b0f3aa504417231fdc81f33a

**Comment by kouloumos** [comment]
Thank you all for the reviews!

A rebase was needed after the merge of https://github.com/bitcoin/bitcoin/pull/26084.
Also force pushed to change the error code from `RPC_WALLET_INSUFFICIENT_FUNDS` to `RPC_WALLET_ERROR` based on @ishaanam's comment.

**Comment by glozow** [comment]
re ACK cc434cb via range-diff. Changes were addressing https://github.com/bitcoin/bitcoin/pull/26024#discussion_r971325299 and https://github.com/bitcoin/bitcoin/pull/26024#discussion_r970651614.

**Comment by achow101** [comment]
ACK cc434cbf583ec8d1b0f3aa504417231fdc81f33a

**Comment by Xekyo** [comment]
post-merge ACK cc434cbf583ec8d1b0f3aa504417231fdc81f33a

Thanks for catching this.

**Pull request merged**
